content/articles/mocking-with-junit-and-mockito-the-why-and-how/index.md

```diff
@@ -0,0 +1,870 @@
+
+3. #### Using JUnit Jupiter's MockitoExtension
+   [JUnit 5](https://junit.org/junit5/) provides an extension - [org.mockito:mockito-JUnit-jupiter](https://search.maven.org/search?q=g:org.mockito%20AND%20a:mockito-junit-jupiter).
+
+   To use the extension:
+  - We add the following dependency to our ```pom.xml``` file.
+    ```xml
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-junit-jupiter</artifactId>
+      <version>3.12.4</version>
+      <scope>test</scope>
+    </dependency>
+    ```
+
+   > This single dependency replaces both the ```mockito core``` and the ```junit``` dependencies defined above.
+
+  - We add ```@ExtendWith(MockitoExtension.class)``` to the test class
+  - We annotate fields that we want to mock with the ```@Mock```  annotation.
+
+   Using the Mockito extension makes our test classes neat and concise.
+    ```java
+    @ExtendWith(MockitoExtension.class)
+    class StudentServiceImpl_StubbingTests {
+        private StudentService studentService;
+        @Mock
+        private StudentRepository studentRepository;
+        @Mock
+        private HostelRepository hostelRepository;
+        private RegistrationDto registrationRequest;
+        private Student student;
+
+        @BeforeEach
+        void setUp() {
+            //we inject the mocks into the studentServiceImpl
+            studentService = new StudentServiceImpl(studentRepository, hostelRepository);
+        }
+        //test cases excluded for the sake of brevity
+    }
+    ```
+
+4. #### Using @InjectMocks with Spring
+   When using Spring, we can inject a mock object into Spring's ```ApplicationContext``` making use of the ```@MockBean``` annotation:
+    ```java
+    @ExtendWith(MockitoExtension.class)
+    class StudentServiceImplTest {
+        @MockBean
+        private StudentRepository studentRepository;
+        @MockBean
+        private HostelRepository hostelRepository;
+        @InjectMocks
+        private StudentService studentService;
+
+        //test cases excluded for the sake of brevity
+    }
+    ```
+
+   > ```@MockBean``` is a SpringBoot annotation. Spring automatically injects the mock object as a replacement for the actual object wherever a dependency of an appropriate bean is required.
+
+### Knowing when to use ```Mockito.mock``` and when to use @MockBean from Springboot.
+
+We use `Mockito.mock` when our test does not rely on the SpringBoot container. Using ```Mockito.mock``` is fast and ensures the isolation of each test component.
+
+We use  ```@MockBean``` from SpringBoot when our test relies on the SpringBoot container, and when we want to mock one of the container's beans. Typical usage of SpringBoot  ```@MockBean``` is in writing controller tests with test classes annotated with ```@WebMvcTest```.
```

Comment:
` ```@MockBean``` `

You can use a single ` `` ` for highlighting keywords or variables. You use these only for code snippets.

Reply:
Do this change throughout the article.